CoreScriptsRoot/CoreScripts/PerformanceStatsManagerScript.lua

```diff
@@ -0,0 +1,184 @@
+    AddButton(displayType, i)
+  end
+end
+
+function OnButtonToggled(toggledDisplayType) 
+  local toggledButton = statsButtonsByType[toggledDisplayType]
+  local selectedState = toggledButton._isSelected
+  selectedState = not selectedState
+  
+  if (selectedState) then 
+    currentDisplayType = toggledDisplayType
+  else
+    currentDisplayType = nil
+  end
+  
+  UpdateButtonSelectedStates()
+  UpdateViewerVisibility()
+end
+
+function UpdateButtonSelectedStates()
+  for i, buttonType in ipairs(StatsUtils.AllStatDisplayTypes) do
+      local button = statsButtonsByType[buttonType]
+      button:SetIsSelected(buttonType == currentDisplayType)
+  end  
+end
+
+function UpdateViewerVisibility()
+  -- If someone is on, show the Viewer.
+  -- FIXME(dbanks)
+  -- Configure with details of the dude currently selected.  
+  if (currentDisplayType == nil) then 
+    statsViewer:SetVisible(false)
+    statsViewer:SetStatsAggregator(nil)
+  else
+    statsViewer:SetStatsDisplayType(currentDisplayType)
+    local aggregatorType = StatsUtils.DisplayTypeToAggregatorType[currentDisplayType]
+    statsViewer:SetStatsAggregator(allStatsAggregators:GetAggregator(aggregatorType))
+    
+    statsViewer:SetVisible(true)
+  end
+end
+
+function AddButton(displayType, index) 
+  -- Configure size and position of button.
+  -- Configure callback behavior to toggle
+  --    button on or off and show/hide viewer.
+  -- Parent button in main screen.
+  local button = statsButtonsByType[displayType]
+  
+  button:SetParent(masterFrame)
+  local aggregatorType = StatsUtils.DisplayTypeToAggregatorType[displayType]
+  button:SetStatsAggregator(
+    allStatsAggregators:GetAggregator(aggregatorType))
+  
+  local fraction = 1.0/StatsUtils.NumButtonTypes
+  local size = UDim2.new(fraction, 0, 0, StatsUtils.ButtonHeight)
+  local position = UDim2.new(fraction * (index - 1), 0, 0, 0)
+  button:SetSizeAndPosition(size, position)
+  
+  button:SetToggleCallbackFunction(OnButtonToggled)
+end
+
+function ConfigureStatViewerInMasterFrame()
+  -- Set up the widget that shows currently selected button.
+  statsViewer:SetParent(masterFrame)
+  
+  local size = UDim2.new(0.5, 0, 0.5, 0)
+  local position = UDim2.new(0.5, 0, 0.25, 0)
+  statsViewer:SetSizeAndPosition(size, position)
+end
+
+function UpdatePerformanceStatsVisibility() 
+  local localPlayer = PlayersService.LocalPlayer
+  
+  local isVisible = (GameSettings.PerformanceStatsVisible and localPlayer ~= nil)
+  if isVisible then 
+    masterFrame.Visible = true
+    masterFrame.Parent = RobloxGuiFolder
+  else
+    masterFrame.Visible = false
+    masterFrame.Parent = nil
+  end
+end
+
+
+--[[ Top Level Code ]]--
+-- If flag is not enabled, bounce.
+if not showPerformanceStatsInGui then 
+	return
+end
+
+-- Set up our GUI.
+ConfigureMasterFrame()
+ConfigureStatButtonsInMasterFrame()
+ConfigureStatViewerInMasterFrame()
+
+PlayersService.PlayerAdded:connect(UpdatePerformanceStatsVisibility)
```

Comment:
I don't think it is. Since we restart the dm when you transition from edit->play and vice versa, PlayerAdded/Removed should never get called in edit mode. But in an online game/testing play mode in studio those functions will be called often (any time any other play joins/leaves the game), but are never necessary after LocalPlayer is initialized once, since it never changes during a play session.